src/silo/preprocessing/metadata_info.cpp

```diff
@@ -0,0 +1,200 @@
+      if (std::find(config_metadata_fields.begin(), config_metadata_fields.end(), pair.first)
+       != config_metadata_fields.end()) {
+         validated_metadata_fields.emplace(pair);
+      }
+   }
+   for (const std::string& name : config_metadata_fields) {
+      if (!validated_metadata_fields.contains(name)) {
+         throw silo::preprocessing::PreprocessingException(fmt::format(
+            "The metadata field {} which is contained in the database config is "
+            "not contained in the input.",
+            name
+         ));
+      }
+   }
+
+   std::string metadata_field_string;
+   for (const auto& [field, select] : validated_metadata_fields) {
+      metadata_field_string += "'" + field + "' with selection '" + select + "',";
+   }
+   SPDLOG_TRACE("Found metadata fields: " + metadata_field_string);
+   return validated_metadata_fields;
+}
+
+void detectInsertionLists(
+   const std::filesystem::path& ndjson_file,
+   std::unordered_map<std::string, std::string>& metadata_fields_to_validate
+) {
+   duckdb::DuckDB db(nullptr);
+   duckdb::Connection connection(db);
+   auto top_level_entries =
+      connection.Query(fmt::format("SELECT * FROM '{}' LIMIT 0", ndjson_file.string()));
+   for (size_t idx = 0; idx < top_level_entries->ColumnCount(); idx++) {
+      const std::string& top_level_entry = top_level_entries->ColumnName(idx);
+      if (top_level_entry == "nucleotideInsertions" || top_level_entry == "aminoAcidInsertions") {
+         auto contained_insertions = connection.Query(
+            fmt::format("SELECT {}.* FROM '{}' LIMIT 0", top_level_entry, ndjson_file.string())
+         );
+         if (contained_insertions->ColumnCount() == 0) {
+            metadata_fields_to_validate[top_level_entry] = "''";
+         }
+         if (contained_insertions->ColumnCount() == 1) {
+            metadata_fields_to_validate[top_level_entry] = fmt::format(
+               "list_string_agg({}.{})", top_level_entry, contained_insertions->ColumnName(0)
+            );
+         }
+
+         std::vector<std::string> list_transforms;
+         for (size_t idx2 = 0; idx2 < contained_insertions->ColumnCount(); idx2++) {
+            const std::string& sequence_name = contained_insertions->ColumnName(idx2);
+            list_transforms.push_back(fmt::format(
+               "list_transform({0}.{1}, x ->'{1}:' || x)", top_level_entry, sequence_name
+            ));
+         }
+         metadata_fields_to_validate[top_level_entry] =
+            "list_string_agg(flatten([" + boost::join(list_transforms, ",") + "]))";
+      }
+   }
+}
+
+}  // namespace
+
+namespace silo::preprocessing {
+
+MetadataInfo MetadataInfo::validateFromMetadataFile(
+   const std::filesystem::path& metadata_file,
+   const silo::config::DatabaseConfig& database_config
+) {
+   if (database_config.schema.metadata.empty()) {
+      throw PreprocessingException("Database config without fields not possible");
+   }
```

Comment:
We can also move this to a more central and earlier location, so that SILO throws an exception already when building the database config.